keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,544 @@
+signature that `fieldManager` is required.
+
+```go
+type ApplyOptions struct {
+  DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
+  Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`
+}
+
+func (ApplyOptions) ToPatchOptions(fieldManager string) PatchOptions
+```
+
+Apply will combine the `fieldManager` argument with `ApplyOptions` to create the
+`PatchOptions`.
+
+Each apply call will be required to provide a fieldmanager name. We will not
+provide a a way for the fieldmanager name to be set for the entire
+clientset. There are a couple reasons for this:
+
+- If a client has multiple code paths where it makes apply requests to the same
+  object, but with different field sets, they must use different field manager
+  names. If they use the same field manager name they will cause fields to be
+  accidentally removed or disowned. This is a potential foot gun we would like to
+  avoid.
+
+- Apply requests always conflict with update requests, even if they were made by
+  the same client with the same field manager name. This is by design. So when a
+  controller migrates from update to apply, it will need to deal with conflicts
+  regardless of what field manager name is used.
+
+### Generated apply configuration types
+
+All fields present in an apply configuration become owned by the applier after
+when the apply request succeeds. Go structs contain zero valued fields which are
+included even if the user never explicitly sets the field. Required boolean
+fields are a good example of fields that would be applied incorrectly using go
+structs, e.g. `ContainerStatus.Ready` (required, not omitempty). Because of this
+we cannot use the existing go structs to represent apply configurations.
+
+<<[UNRESOLVED @jpbetz @jennybuckley ]>> 
+Finalize which alternative to use based on developer feedback. See the
+[Alternatives](#alternatives) for a complete list, but are currently focusing on
+the two below alternatives. We are working with the Kubebuilder community to
+gather feedback on what developers prefer.
+<<[/UNRESOLVED]>>
+
+#### Alternative 1: Genreated structs where all fields are pointers
+
+Example usage:
+
+```go
+&appsv1apply.Deployment{
+  Name: &pointer.StringPtr("nginx-deployment"),
+  Spec: &appsv1apply.DeploymentSpec{
+    Replicas: &pointer.Int32Ptr(0),
+    Template: &v1apply.PodTemplate{
+      Spec: &v1apply.PodSpec{
+        Containers: []v1.Containers{
+          {
+            Name: &pointer.StringPtr("nginx"),
+            Image: &pointer.StringPtr("nginx:latest"),
+          },
+        }
+      },
+    },
+  },
+}
+```
+
+For this approach, developers need to use a library like
+https://github.com/kubernetes/utils/blob/master/pointer/pointer.go
+to inline primitive literals.
+
+#### Alternative 2: Generated "builders"
+
+Example usage:
+
+```go
+&appsv1apply.Deployment().
+  SetObjectMeta(&metav1apply.ObjectMeta().
```

Comment:
Do you really need the "Set" prefix for everything?

Reply:
Golang libraries uses `With` instead of `Set` a lot which I prefer. But Set is a known thing inside the Kubernetes types so far so feels more uniform.

Reply:
After a few discussions, I've switched the example to not prefix the builder name with "Set", so now we just have `ObjectMeta()` (instead of `SetObjectMeta()`). I've also added a `<<[UNRESOLVED` to this it doesn't block us from merging this KEP as provisional.